server/text_generation_server/utils/tokens.py

```diff
@@ -350,6 +394,19 @@ def __call__(
         else:
             speculative_ids = None
 
+        # advance the grammar state
+        if self.grammar_processor is not None:
+            for i in range(len(self.fsm_grammar_states)):
+                try:
+                    self.fsm_grammar_states[i] = self.grammar_processor.advance(
+                        next_ids[i].item(), self.fsm_grammar_states[i], self.grammars[i]
+                    )
```

Comment:
The whole point of HeteregenousProcessor is to processor all the tokens at once, without making any CPU calls.

This `.item()` defeats the purpose.

I don't have any good ideas aside from moving this bit to the CPU loop in `causal_lm`/ `flash_causal_lm`.
Which is probably error prone.

@OlivierDehaene Do you have better suggestions ?